server/blob/blob-cassandra/src/main/java/org/apache/james/blob/cassandra/cache/CassandraDumbBlobStoreCache.java

```diff
@@ -0,0 +1,119 @@
+package org.apache.james.blob.cassandra.cache;
+
+import static com.datastax.driver.core.querybuilder.QueryBuilder.bindMarker;
+import static com.datastax.driver.core.querybuilder.QueryBuilder.eq;
+import static org.apache.james.blob.cassandra.BlobTables.BucketBlobTable.ID;
+import static org.apache.james.blob.cassandra.BlobTables.DumbBlobCache.DATA;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.ByteBuffer;
+
+import javax.inject.Inject;
+
+import org.apache.james.backends.cassandra.utils.CassandraAsyncExecutor;
+import org.apache.james.blob.api.BlobId;
+import org.apache.james.blob.cassandra.BlobTables;
+import org.apache.james.blob.cassandra.utils.DataChunker;
+import org.reactivestreams.Publisher;
+
+import com.datastax.driver.core.PreparedStatement;
+import com.datastax.driver.core.Session;
+import com.datastax.driver.core.querybuilder.QueryBuilder;
+import com.google.common.base.Preconditions;
+
+import reactor.core.publisher.Mono;
+
+public class CassandraDumbBlobStoreCache implements DumbBlobStoreCache {
+
+    private final DataChunker dataChunker;
+    private final CassandraAsyncExecutor cassandraAsyncExecutor;
+    private final PreparedStatement insert;
+    private final PreparedStatement select;
+    private final PreparedStatement delete;
+    private final CassandraCacheConfiguration cacheConfiguration;
+
+    @Inject
+    public CassandraDumbBlobStoreCache(Session session, CassandraCacheConfiguration cacheConfiguration) {
+        this.dataChunker = new DataChunker();
+        this.cassandraAsyncExecutor = new CassandraAsyncExecutor(session);
+        this.insert = prepareInsert(session);
+        this.select = prepareSelect(session);
+        this.delete = prepareDelete(session);
+        this.cacheConfiguration = cacheConfiguration;
+    }
+
+    private PreparedStatement prepareDelete(Session session) {
+        return session.prepare(
+            QueryBuilder.delete().
+                from(BlobTables.BucketBlobTable.TABLE_NAME)
+                .where(eq(ID, bindMarker(ID))));
+    }
+
+    private PreparedStatement prepareSelect(Session session) {
+        return session.prepare(
+            QueryBuilder.select()
+                .from(BlobTables.BucketBlobTable.TABLE_NAME)
+                .where(eq(ID, bindMarker(ID))));
+    }
+
+    private PreparedStatement prepareInsert(Session session) {
+        return session.prepare(
+            QueryBuilder.insertInto(BlobTables.DumbBlobCache.TABLE_NAME)
+                .value(ID, bindMarker(ID))
+                .value(DATA, bindMarker(DATA)));
+    }
+
+    private boolean isAValidStream(InputStream bufferedData) throws IOException {
```

Comment:
InputStream should always be a buffered inputStream for this to work. Please enforce it via a type.

Reply:
not true, it works with a FileInputStream and a ByteInputStream for example.
There's no typing that can save you here.
The only safe move is to wrap the InputStream into a BufferedInputStream locally, event if it's wasteful

Reply:
Then what your suggestion?

Reply:
> The only safe move is to wrap the InputStream into a BufferedInputStream locally, event if it's wasteful

Reply:
> The only safe move is to wrap the InputStream into a BufferedInputStream locally, event if it's wasteful

But once you now if the original input stream is "valid" or not, you need to keep reading the buffered one to not loose content.

> not true, it works with a FileInputStream and a ByteInputStream for example.

Well, we don't deal with such input types here: we only call it with a BufferedInputStream.

We don't try to be generic.

By passing a `BufferedInputStream` as an argument we would ensure that the caller don't forget to wrap it's input stream and buy (limited) extra safety.

> Then what your suggestion?

`private boolean isAValidStream(BufferedInputStream bufferedData)`

As a method signature.

Reply:
Thank you.